# Exercise 2.A Fundamentals of pandas

In [25]:
# Importing NumPy and giving it the nickname np
import numpy as np

# Importing pandas and giving it the nickname pd
import pandas as pd

## Series

In [26]:
# Creating a list object
aa = [1, 2, 3, 4]

# Creating a tuple object
ab = (5, 6, 7, 8)

# Creating a set object
ac = {9, 10, 11, 12}

# Creating a dictionary object
ad = {
    'first': 13,
    'second': 14,
    'third': 15,
    'fourth': 16
}

In [27]:
# Creating a Series from list aa
ser1 = pd.Series(aa)

# Creating a Series from tuple ab
ser2 = pd.Series(ab)

# Creating a Series from set ac
# Initially did ser3=pd.Series(ac) but it gave me a TypeError since sets are unordered
# Converting set to list
ser3 = pd.Series(list(ac))

# Creating a Series from dictionary ad
ser4 = pd.Series(ad)

# Printing each Series
print(ser1)
print()

print(ser2)
print()

print(ser3)
print()

print(ser4)

0    1
1    2
2    3
3    4
dtype: int64

0    5
1    6
2    7
3    8
dtype: int64

0     9
1    10
2    11
3    12
dtype: int64

first     13
second    14
third     15
fourth    16
dtype: int64


In [28]:
# Creating a Series using literal values
ser5 = pd.Series(17, 18, 19, 20)

# Display the Series
print(ser5)

# TypeError: Index(...) must be called with a collection of some kind, 18 was passed
# The problem is that pandas thought 18 was supposed to be the index, but it was not given as a list or collection.

TypeError: Index(...) must be called with a collection of some kind, 18 was passed

## DataFrames from Series

In [29]:
# Creating a variable df1 and storing a DataFrame in it
# Using data from ser1 to create a table
df1 = pd.DataFrame(ser1)

# Displaying df1
df1

,0
0,1
1,2
2,3
3,4


In [30]:
# Inserting ser2 into column position 1
df1.insert(1, 'Column_2', ser2)

# Inserting ser3 into column position 2
df1.insert(2, 'Column_3', ser3)

# Display updated DataFrame
df1

# insert(index position, column name, data)

,0,Column_2,Column_3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


In [31]:
# Adding ser4 as a new column
df1['Column_4'] = ser4

# Display the updated DataFrame
df1

,0,Column_2,Column_3,Column_4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN


ser4 behaved differently because it was created from a dictionary, which uses custom index labels ('first', 'second', 'third', 'fourth') instead of numbered indexes (0,1,2,3).

Since df1 uses numbered indexes and ser4 uses text labels, pandas could not match the rows correctly. Because of this mismatch, pandas filled the new column with NaN values.

In [32]:
# .loc is creating row 4 and adding the specified values inside
df1.loc[4] = [5, 9, 12, 0]

# Display updated DataFrame
df1

,0,Column_2,Column_3,Column_4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN
4,5,9,12,0.0


## Reading Data into a DataFrame

In [33]:
# Reading sakila_example.csv and turning it into a DataFrame
df2 = pd.read_csv('sakila_example.csv')

# Displaying the DataFrame
df2

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


In [34]:
# Displaying the first 5 rows
df2.head()

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


In [35]:
# Displaying the last 10 rows
df2.tail(10)

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


In [36]:
# Displaying summary information about the DataFrame
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   title        262 non-null    str  
 1   description  262 non-null    str  
 2   category     262 non-null    str  
 3   language     262 non-null    str  
 4   actor_name   262 non-null    str  
dtypes: str(5)
memory usage: 10.4 KB


### Facts about the dataset from info()

- The dataset contains 262 rows (entries) and 5 columns.
- The columns are: title, description, category, language, and actor_name.
- All columns contain 262 non-null values, meaning there are no missing values in the dataset.
- All columns are stored as string (str) data types, meaning the dataset mainly contains text data.
- The dataset uses approximately 10.4 KB of memory, which is relatively small.

In [37]:
# Displaying column statistics
df2.describe()

,title,description,category,language,actor_name
count,262,262,262,262,262
unique,46,46,15,1,145
top,ARABIA DOGMA,A Touching Epistle of a Madman And a Mad Cow w...,Horror,English,OPRAH KILMER
freq,12,12,55,262,8


### Insights from describe()

- The dataset contains 262 records for every column, confirming there are no missing values.
- There are 46 unique movie titles and 46 unique descriptions, meaning some movies appear multiple times.
- There are 15 unique movie categories, and Horror is the most common category, appearing 55 times.
- The language column contains only 1 unique value, meaning all movies are in English.
- There are 145 unique actor names, with OPRAH KILMER appearing most frequently (8 times).

### Less useful summarizations

- Since the dataset contains mostly text/string data, numerical statistics like averages or totals are not available.
- The summary is useful for identifying repeated values and categories, but not useful for mathematical analysis.

In [38]:
# Reading northwind_example.csv and turning it into a DataFrame
df3 = pd.read_csv('northwind_example.csv')

# Displaying the DataFrame
df3

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


In [39]:
# Displaying column statistics
df3.describe()

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


### Insights from describe() for df3

- The dataset contains 500 records for all numeric columns, meaning there are no missing numeric values.
- The average (mean) UnitPrice is approximately 23.37, while the average Quantity ordered is about 24 items.
- Product prices vary significantly, ranging from 2.00 to 210.80.
- Order quantities range from 1 item to 120 items, showing that some orders are much larger than others.
- The median (50%) quantity ordered is 20, meaning half of the orders were above 20 items, and half were below.

### How this summary differs from the previous example

- This dataset contains numeric data, so describe()` provides mathematical statistics such as mean, standard deviation, minimum, maximum, and percentiles.
- The previous dataset (df2) contained text/string data, so describe() focused on unique values, most common values (top), and frequency (freq) instead.
- Numeric datasets allow for deeper mathematical analysis, while text datasets are better for category and frequency analysis.

## Coding a DataFrame

In [40]:
# Creating a dictionary that stores sample student information
test_data = {
    'ID': [1, 2, 3, 4, 5, 6, 7, 8],
    'Name': ['Alicia', 'Bob', 'Charlie', 'Dominic', 'Eve', 'Frank', 'Grace', 'Heidi'],
    'Age': [25, np.nan, 22, 19, 31, 35, np.nan, 18],
    'City': ['NY', 'LA', 'Chicago', 'Houston', 'Phoenix', 'Boston', 'Austin', 'San Diego'],
    'Score': [85, 92, -1, 65, 78, np.nan, 55, 90]
}

In [41]:
# Creating a DataFrame from the dictionary
df4 = pd.DataFrame(test_data)

# Displaying the DataFrame
df4

,ID,Name,Age,City,Score
0,1,Alicia,25.0,NY,85.0
1,2,Bob,NaN,LA,92.0
2,3,Charlie,22.0,Chicago,-1.0
3,4,Dominic,19.0,Houston,65.0
4,5,Eve,31.0,Phoenix,78.0
5,6,Frank,35.0,Boston,NaN
6,7,Grace,NaN,Austin,55.0
7,8,Heidi,18.0,San Diego,90.0


In [43]:
# Exporting df4 into a CSV file
df4.to_csv('test_data.csv', index=False)
# index=false means the row numbers get saved into the csv too